### ISSUE TITLE
Error MT4109: Failed to compile the generated registrar code

### ISSUE THREAD

**Issue by mehulsantifm** [xamarin/xamarin-macios#19122]
### Steps to Reproduce

1. update Mac OS to 14
2. Xcode to 15
3. on iPhone XR 17.0.2 building Xamarin app ended into this error

### Expected Behaviour
Need to build and able to run the app

### Actual Behaviour
Error - Error MT4109: Failed to compile the generated registrar code

### Environment

Visual Studio Community 2022 for Mac
Version 17.6.4 (build 472)
Installation UUID: 7c66dcd3-cfec-41c2-b0c5-3a7127c9d564

Runtime
.NET 7.0.3 (64-bit)
Architecture: Arm64
Microsoft.macOS.Sdk 13.1.1007; git-rev-head:8afca776a0a96613dfb7200e0917bb57f9ed5583; git-branch:release/7.0.1xx-xcode14.2

Roslyn (Language Service)
4.6.0-3.23180.6+99e956e42697a6dd886d1e12478ea2b27cceacfa

NuGet
Version: 6.4.0.117

.NET SDK (Arm64)
SDK: /usr/local/share/dotnet/sdk/7.0.308/Sdks
SDK Versions:
	7.0.308
	7.0.307
	7.0.302
	6.0.414
	6.0.413
	6.0.408
MSBuild SDKs: /Applications/Visual Studio.app/Contents/MonoBundle/MSBuild/Current/bin/Sdks

.NET Runtime (Arm64)
Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	7.0.11
	7.0.10
	7.0.5
	6.0.22
	6.0.21
	6.0.16

Xamarin.Profiler
Version: 1.8.0.49
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

Updater
Version: 11

Apple Developer Tools
Xcode: 15.0 22265
Build: 15A240d

Xamarin.Mac
Not Installed

Xamarin.iOS
Version: 16.4.0.6 Visual Studio Community
Hash: 97731c92c
Branch: xcode14.3
Build date: 2023-04-11 22:38:36-0400

Xamarin.Android
Version: 13.2.1.2 (Visual Studio Community)
Commit: xamarin-android/d17-5/a8a26c7
Android SDK: /Users/mehulsant/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		12.0 (API level 31)
		9.0  (API level 28)
		13.0 (API level 33)

SDK Command-line Tools Version: 7.0
SDK Platform Tools Version: 34.0.1
SDK Build Tools Version: 32.0.0

Build Information: 
Mono: d9a6e87
Java.Interop: xamarin/java.interop/d17-5@149d70fe
SQLite: xamarin/sqlite/3.40.1@68c69d8
Xamarin.Android Tools: xamarin/xamarin-android-tools/d17-5@ca1552d

Microsoft Build of OpenJDK
Java SDK: /Library/Java/JavaVirtualMachines/microsoft-11.jdk
11.0.16.1
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

Eclipse Temurin JDK
Java SDK: /Library/Java/JavaVirtualMachines/temurin-8.jdk
1.8.0.302
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

Android SDK Manager
Version: 17.6.0.50
Hash: a715dca
Branch: HEAD
Build date: 2023-09-07 02:05:26 UTC

Android Device Manager
Version: 0.0.0.1309
Hash: 06e3e77
Branch: HEAD
Build date: 2023-09-07 02:05:26 UTC

Xamarin Designer
Version: 17.6.3.9
Hash: 2648399ae8
Branch: remotes/origin/d17-6
Build date: 2023-09-07 02:05:20 UTC

Build Information
Release ID: 1706040472
Git revision: 0b8c2cb9f01ef14a2b07ff4ea047268c8756fee6
Build date: 2023-09-07 02:03:50+00
Build branch: release-17.6
Build lane: release-17.6

Operating System
Mac OS X 14.0.0
Darwin 23.0.0 Darwin Kernel Version 23.0.0
    Fri Sep 15 14:42:57 PDT 2023
    root:xnu-10002.1.13~1/RELEASE_ARM64_T8112 arm64

### Build Logs

MTOUCH: Error MT4109: Failed to compile the generated registrar code. Please file a bug report at https://github.com/xamarin/xamarin-macios/issues/new (MT4109)

**Comment by CristianValencia** [comment]
Also got this; Initially I didn't want to upgrade the whole trinity of chaos (VS for Mac + XCode + OS) but I did and unfortunately now I'm getting this error while trying to deploy. 

Visual Studio for Mac version is 17.6.4 (build 472). 
XCode version is 15.0
macOS Sonoma 14.0

Also, all my iOS device simulators are gone, so basically I can't deploy either way.

**Comment by pulmuone** [comment]
same issue

**Comment by DaoChiMinh** [comment]
same issue

**Comment by vyn203** [comment]
I have the same problem along with the **Lower the "Deployment Target"** issue for simulators. Can't revert back to an older version of Xcode since Sonoma doesn't support anything before 15.0

I did find a temporary fix by adding a trailing slash to the Apple SDK location **/Applications/Xcode.app to /Applications/Xcode.app/** but the issue still occurs after restarting VS.

**Comment by vyn203** [comment]
https://github.com/xamarin/xamarin-macios/issues/18964 this solved the issues for me

**Comment by vyn203** [comment]
https://github.com/xamarin/xamarin-macios/issues/18964 - In the first comment of this issue, it is suggested to download the latest version of Xamarin.ios and Xamarin.mac. The packages are linked. I manually downloaded both onto my computer, and from there on, all of my simulators, linker issues, and errors were solved. Make sure to include the trailing dash for the Apple SDK location. There wasn't any nuget packages that needed to be updated for my case.
